sdk/dotnet/Pulumi/Resources/ComponentResourceOptions.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright 2016-2018, Pulumi Corporation
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Pulumi
+{
+    /// <summary>
+    /// A bag of optional settings that control a <see cref="ComponentResource"/>'s behavior.
+    /// </summary>
+    public class ComponentResourceOptions : ResourceOptions
+    {
+        private List<ProviderResource>? _providers;
+
+        /// <summary>
+        /// An optional set of providers to use for child resources.
+        ///
+        /// Note: do not provide both <see cref="ResourceOptions.Provider"/> and <see
+        /// cref="Providers"/>.
+        /// </summary>
+        public List<ProviderResource> Providers

```

Comment:
We should consider using `Collection<ProviderResource>` here, without the property setter, to following the Framework Design Guidelines for collections: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections

> **X DO NOT** use `ArrayList` or `List<T>` in public APIs.

> **X DO NOT** provide settable collection properties.

> **✓ DO** use `Collection<T>` or a subclass of `Collection<T>` for properties or return values representing read/write collections.

Reply:
> We should consider using Collection<ProviderResource> here, without the property setter, to following the Framework Design Guidelines for collections:

Reply:
> As well as
> 
> ```
> new ResourceOptions { Providers = { p1, p2 } }
> ```

Note that these guidelines don't preclude this from working. If the property was defined as:

```c#
public Collection<ProviderResource> Providers { get; } = new Collection<ProviderResource>();
```

You can still use collection initializers, e.g.:

```c#
new ResourceOptions { Providers = { p1, p2 } }
```

Which is compiled as:

```c#
var options = new ResourceOptions();
options.Providers.Add(p1);
options.Providers.Add(p2);
```

Reply:
> ote that these guidelines don't preclude this from working. If the property was defined as:

It does, because it says: `X DO NOT provide settable collection properties.`

However, we want to support: `new SomeResource { InputProps = OtherResource.OutputProps }`. We need a settable collection in this case.

The design here is to enable both forms.  i.e. taking values wholesale and providing them (i.e. setting the entire list), and also taking individual values and adding them.

Reply:
OK, fair enough :)